sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezZPageHandler.java

```diff
@@ -0,0 +1,605 @@
+      out.print("</tr>");
+      if (calendar.get(Calendar.DAY_OF_YEAR) != lastEntryDayOfYear) {
+        lastEntryDayOfYear = calendar.get(Calendar.DAY_OF_YEAR);
+      }
+      lastEpochNanos = event.getEpochNanos();
+      zebraStripe = !zebraStripe;
+    }
+    formatter.format(
+        "<tr style=\"background-color: %s;\"><td></td><td class=\"border-left-dark\">"
+            + "</td><td class=\"border-left-dark\"><pre class=\"no-margin wrap-text\">",
+        zebraStripe ? ZEBRA_STRIPE_COLOR : "#fff");
+    Status status = span.getStatus();
+    if (status != null) {
+      formatter.format("%s | ", htmlEscaper().escape(status.toString()));
+    }
+    formatter.format("%s</pre></td>", htmlEscaper().escape(renderAttributes(span.getAttributes())));
+    zebraStripe = !zebraStripe;
+    return zebraStripe;
+  }
+
+  private static void emitSingleEvent(
+      PrintStream out,
+      Formatter formatter,
+      Event event,
+      Calendar calendar,
+      int lastEntryDayOfYear,
+      long lastEpochNanos) {
+    if (calendar.get(Calendar.DAY_OF_YEAR) == lastEntryDayOfYear) {
+      out.print("<td class=\"align-right\"><pre class=\"no-margin wrap-text\">");
+    } else {
+      formatter.format(
+          "<td class=\"align-right\"><pre class=\"no-margin wrap-text\">%04d/%02d/%02d-",
+          calendar.get(Calendar.YEAR),
+          calendar.get(Calendar.MONTH) + 1,
+          calendar.get(Calendar.DAY_OF_MONTH));
+    }
+
+    // Special printing so that durations smaller than one second
+    // are left padded with blanks instead of '0' characters.
+    // E.g.,
+    //        Number                  Printout
+    //        ---------------------------------
+    //        0.000534                  .   534
+    //        1.000534                 1.000534
+    long deltaMicros = TimeUnit.NANOSECONDS.toMicros(event.getEpochNanos() - lastEpochNanos);
+    String deltaString;
+    if (deltaMicros >= 1000000) {
+      deltaString = String.format("%.6f", (deltaMicros / 1000000.0));
+    } else {
+      deltaString = String.format("%1s.%6d", "", deltaMicros);
+    }
+
+    long microsField = TimeUnit.NANOSECONDS.toMicros(event.getEpochNanos());
+    formatter.format(
+        "%02d:%02d:%02d.%06d</pre></td> "
+            + "<td class=\"border-left-dark\"><pre class=\"no-margin wrap-text\">%s</pre></td>"
+            + "<td class=\"border-left-dark\"><pre class=\"no-margin wrap-text\">%s</pre></td>",
+        calendar.get(Calendar.HOUR_OF_DAY),
+        calendar.get(Calendar.MINUTE),
+        calendar.get(Calendar.SECOND),
+        microsField,
+        deltaString,
+        htmlEscaper().escape(renderEvent(event)));
+  }
+
+  private static String renderAttributes(ReadableAttributes attributes) {
+    final StringBuilder stringBuilder = new StringBuilder();
+    stringBuilder.append("Attributes:{");
+    final AtomicBoolean first = new AtomicBoolean(true);
```

Comment:
You can add `first` as a boolean field of the anonymous `KeyValueConsumer` instead of using `AtomicBoolean` though it could be a bit more readable to separate out a named class which you pass the `StringBuilder` in explicitly

Reply:
The `first` field allows the overridden `consume` function to print something different for the first of all `KeyValueConsumers`. Wouldn't making it a boolean field within each `KeyValueConsumer` defeat its purpose?

Reply:
Nah, there's only one `KeyValueConsumer` (there's only one execution of `new KeyValueConsumer` - using a field means accessing this variable doesn't use the thread-safe atomicboolean, which isn't needed here.

Reply:
I see. I moved it into the `KeyValueConsumer`.